common/raw_hashtable.h

```diff
@@ -709,26 +727,74 @@ ViewImpl<InputKeyT, InputValueT, InputKeyContextT>::ForEachEntry(
 }
 
 template <typename InputKeyT, typename InputValueT, typename InputKeyContextT>
-auto ViewImpl<InputKeyT, InputValueT, InputKeyContextT>::CountProbedKeys(
-    KeyContextT key_context) const -> ssize_t {
+auto ViewImpl<InputKeyT, InputValueT, InputKeyContextT>::GetMetrics(
+    KeyContextT key_context) const -> Metrics {
   uint8_t* local_metadata = metadata();
   EntryT* local_entries = entries();
   ssize_t local_size = alloc_size_;
-  ssize_t count = 0;
+
+  Metrics metrics;
+
+  // Compute the ones we can directly.
+  metrics.deleted_count = llvm::count(
+      llvm::ArrayRef(local_metadata, local_size), MetadataGroup::Deleted);
+  metrics.storage_bytes = AllocByteSize(local_size);
+
+  // We want to process present slots specially to collect metrics on their
+  // probing behavior.
   for (ssize_t group_index = 0; group_index < local_size;
        group_index += GroupSize) {
     auto g = MetadataGroup::Load(local_metadata, group_index);
     auto present_matched_range = g.MatchPresent();
     for (ssize_t byte_index : present_matched_range) {
+      ++metrics.key_count;
       ssize_t index = group_index + byte_index;
       HashCode hash =
           key_context.HashKey(local_entries[index].key(), ComputeSeed());
-      ssize_t hash_index = hash.ExtractIndexAndTag<7>().first &
-                           ComputeProbeMaskFromSize(local_size);
-      count += static_cast<ssize_t>(hash_index != group_index);
+      auto [hash_index, tag] = hash.ExtractIndexAndTag<7>();
+      ProbeSequence s(hash_index, local_size);
+      metrics.probed_key_count +=
+          static_cast<ssize_t>(s.index() != group_index);
+
+      // For each probed key, go through the probe sequence to find both the
+      // probe distance and how many comparisons are required.
+      ssize_t distance = 0;
+      ssize_t compares = 0;
+      for (;; s.Next()) {
+        ssize_t probe_index = s.index();
+        MetadataGroup probe_g =
+            MetadataGroup::Load(local_metadata, probe_index);
+        auto metadata_matched_range = probe_g.Match(tag);
+        if (probe_index != group_index) {
+          compares += std::distance(metadata_matched_range.begin(),
+                                    metadata_matched_range.end());
+          distance += 1;
+          continue;
+        }
+
+        CARBON_CHECK(!metadata_matched_range.empty());
+        for (ssize_t match_index : metadata_matched_range) {
+          if (match_index >= byte_index) {
+            // Note we only count the compares that will *fail* as part of
+            // probing. The last successful compare isn't interesting, it is
+            // always needed.
+            break;
+          }
+          compares += 1;
+        }
+        break;
+      }
```

Comment:
I find this loop structure a bit awkward, especially given that the loop body ends in `break`. Maybe we could split this into work for the groups we probed past and work we do for the group actually containing the key?
```suggestion
      for (; s.index() != group_index; s.Next()) {
        auto metadata_matched_range =
            MetadataGroup::Load(local_metadata, s.index()).Match(tag);
        auto metadata_matched_range = probe_g.Match(tag);
        compares += std::distance(metadata_matched_range.begin(),
                                  metadata_matched_range.end());
        distance += 1;
      }

      auto metadata_matched_range =
          MetadataGroup::Load(local_metadata, s.index()).Match(tag);
      CARBON_CHECK(!metadata_matched_range.empty());
      for (ssize_t match_index : metadata_matched_range) {
        if (match_index >= byte_index) {
          // Note we only count the compares that will *fail* as part of
          // probing. The last successful compare isn't interesting, it is
          // always needed.
          break;
        }
        compares += 1;
      }
```